component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -336,6 +327,136 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
+        UserRealm userRealm = getUserRealm(username);
+        try {
+            if (userRealm != null) {
+                UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+                if (userStoreManager != null) {
+                    String savedOTPString = userStoreManager
+                            .getUserClaimValue(tenantAwareUsername,
+                                    EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM, null);
+                    if (StringUtils.isNotEmpty(savedOTPString)) {
+                        savedOTPs = savedOTPString.split(",");
+                    }
+                }
+            }
+            // Check whether there is any backup OTPs and return.
+            if (ArrayUtils.isEmpty(savedOTPs)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("The claim " + EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM + " does " +
+                            "not contain any values.");
+                }
+                throw new AuthenticationFailedException("The claim " +
+                        EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM +
+                        " does not contain any values", authenticatedUser);
+            }
+            if (isBackUpCodeValid(savedOTPs, userToken)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Found saved backup Email OTP for user :" + authenticatedUser);
+                }
+                context.setSubject(authenticatedUser);
+                savedOTPs = (String[]) ArrayUtils.removeElement(savedOTPs, userToken);
+                userRealm.getUserStoreManager().setUserClaimValue(tenantAwareUsername,
+                        EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM, String.join(",", savedOTPs),
+                        null);
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("User entered OTP :" + userToken + " does not match with any of the saved " +
+                            "backup codes.");
+                }
+                context.setProperty(EmailOTPAuthenticatorConstants.CODE_MISMATCH, true);
+                throw new AuthenticationFailedException("Verification Error due to Code " + userToken + " mismatch.",
+                        authenticatedUser);
+            }
+        } catch (UserStoreException e) {
+            throw new AuthenticationFailedException("Cannot find the user claim for OTP list for user : " +
+                    authenticatedUser, e);
+        }
+    }
+
+    /**
+     * Validates the usertoken from the saved backup otp codes.
+     *
+     * @param savedOTPs Array of saveOTPs.
+     * @param userToken Usertoken.
+     * @return True if the backup is valid, else returns false.
+     */
+    private boolean isBackUpCodeValid(String[] savedOTPs, String userToken) {
+
+        if (StringUtils.isEmpty(userToken)) {
+            return false;
```

Comment:
add a debug log

Reply:
we can remove this if condition since we already check whether the usertoken is empty and throw error if it is empty